warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,278 @@
+
+        if (!(capabilities.containsKey(props.getProperty(HttpWarpScriptExtension.HTTP_CAPABILITY)))) {
+          throw new WarpScriptException("Capability " + HttpWarpScriptExtension.HTTP_CAPABILITY + " is required by function " + getName());
+        }
+      }
+
+    } else if (!stack.isAuthenticated()) {
+      throw new WarpScriptException(getName() + " requires the stack to be authenticated.");
+    }
+
+    //
+    // Retrieve arguments
+    //
+
+    String method = (String) formattedArgs.get(METHOD);
+    Map<Object, Object> properties = (Map) formattedArgs.get(HEADER);
+    String body = (String) formattedArgs.get(BODY);
+
+    //
+    // Check URL
+    //
+
+    URL url = null;
+    try {
+      url = new URL((String) formattedArgs.get(URL));
+    } catch (MalformedURLException mue) {
+      throw new WarpScriptException(getName() + " encountered an invalid URL.", mue);
+    }
+
+    //
+    // Check that we do not exceed the limits
+    //
+
+    // Get the current counters in the stack and initialize them if not present.
+    AtomicLong urlCount;
+    AtomicLong downloadSize;
+
+    try {
+      stackCountersLock.lockInterruptibly();
+
+      Object ufCount = stack.getAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_COUNT);
+      Object ufSize = stack.getAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE);
+
+      if (null == ufCount || null == ufSize) {
+        urlCount = new AtomicLong();
+        downloadSize = new AtomicLong();
+        stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_COUNT, urlCount);
+        stack.setAttribute(HttpWarpScriptExtension.ATTRIBUTE_HTTP_SIZE, downloadSize);
+      } else {
+        urlCount = (AtomicLong) ufCount;
+        downloadSize = (AtomicLong) ufSize;
+      }
+    } catch (InterruptedException ie) {
+      throw new WarpScriptException(getName() + " thread has been interrupted", ie);
+    } finally {
+      if (stackCountersLock.isHeldByCurrentThread()) {
+        stackCountersLock.unlock();
+      }
+    }
+
+    if (urlCount.get() + 1 > (long) HttpWarpScriptExtension.getLongAttribute(stack, HttpWarpScriptExtension.ATTRIBUTE_HTTP_LIMIT)) {
+      throw new WarpScriptException(getName() + " is limited to " + HttpWarpScriptExtension.getLongAttribute(stack, HttpWarpScriptExtension.ATTRIBUTE_HTTP_LIMIT) + " calls.");
+    }
+
+    // Recheck the count here in case of concurrent runs
+    if (urlCount.addAndGet(1) > (long) HttpWarpScriptExtension.getLongAttribute(stack, HttpWarpScriptExtension.ATTRIBUTE_HTTP_LIMIT)) {
+      throw new WarpScriptException(getName() + " is limited to " + HttpWarpScriptExtension.getLongAttribute(stack, HttpWarpScriptExtension.ATTRIBUTE_HTTP_LIMIT) + " calls.");
+    }
+
+    HttpURLConnection conn = null;
+
+    List<Object> res = new ArrayList<Object>();
+
+    try {
+      conn = (HttpURLConnection) url.openConnection();
+
+      if (null != url.getUserInfo()) {
```

Comment:
I think it it better to disallow user infos in the URL and to provide a macro to encode basic authentication which can then be passed in the headers.